golang/cosmos/x/vstorage/keeper/keeper.go

```diff
@@ -23,14 +24,82 @@ type StreamCell struct {
 	Values      []string `json:"values"`
 }
 
+type Change struct {
+	Path         string
+	Value        string
+	OldValue     string
+	LegacyEvents bool
+}
+
+type ChangeManager interface {
+	Track(ctx sdk.Context, k Keeper, path, value string, isLegacy bool)
+	EmitEvents(ctx sdk.Context, k Keeper)
+	Rollback(ctx sdk.Context)
+}
+
+type BatchingChangeManager struct {
+	changes map[string]*Change
+}
+
+var _ ChangeManager = (*BatchingChangeManager)(nil)
+
 // Keeper maintains the link to data storage and exposes getter/setter methods
 // for the various parts of the state machine
 type Keeper struct {
-	storeKey sdk.StoreKey
+	changeManager ChangeManager
+	storeKey      sdk.StoreKey
+}
+
+func (bcm BatchingChangeManager) Track(ctx sdk.Context, k Keeper, path, value string, isLegacy bool) {
+	if change, ok := bcm.changes[path]; ok {
+		change.Value = value
+		if isLegacy {
+			change.LegacyEvents = true
+		}
+		return
+	}
+	bcm.changes[path] = &Change{
+		Path:         path,
+		Value:        value,
+		OldValue:     k.GetData(ctx, path),
+		LegacyEvents: isLegacy,
+	}
+}
+
+func (bcm *BatchingChangeManager) Rollback(ctx sdk.Context) {
+	bcm.changes = make(map[string]*Change)
+}
+
+func (bcm BatchingChangeManager) EmitEvents(ctx sdk.Context, k Keeper) {
+	changes := bcm.changes
+	if len(changes) == 0 {
+		return
+	}
+
+	// Deterministic order.
+	sortedPaths := make([]string, 0, len(changes))
+	for path := range changes {
+		sortedPaths = append(sortedPaths, path)
+	}
+	sort.Strings(sortedPaths)
+
+	for _, path := range sortedPaths {
+		change := bcm.changes[path]
+		k.EmitChange(ctx, change)
+	}
+}
```

Comment:
Do you want to clear the change cache when you're done emitting? I see that you explicitly call `Rollback()` in `FlushChangeEvents()`, so it's a matter of personal preference, but at least comment whether the method should or shouldn't clear the cache.